### PR TITLE
Support file split in `NetCDFOutputWriter` when `max_filesize` is exceeded

### PR DESCRIPTION
**Pull request by josuemtzmo**
This PR aims to allow the user to split netCDF files into smaller files when the `filesize` exceeds the `max_filesize`.  This addresses https://github.com/CliMA/Oceananigans.jl/issues/2967

### CODE REVIEW

**Comment by josuemtzmo** [comment]
Although the new `test_netcdf_file_splitting` is working, I'm currently having issues in with the `test_netcdf_time_averaging`, where the test `test_netcdf_time_averaging` fails.
https://github.com/CliMA/Oceananigans.jl/blob/3bb62a647a55a7dadf5f37331321bf0020a78c4d/test/test_netcdf_output_writer.jl#L683
I've tested the PR & the `main` upstream branch in my computer and I have the same error:
```
Stacktrace:
 [1] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/Test/src/Test.jl:672 [inlined]
 [2] test_netcdf_time_averaging(arch::CPU)
   @ Main ~/github/Oceananigans.jl/test/test_netcdf_output_writer.jl:727
 [3] macro expansion
   @ ~/github/Oceananigans.jl/test/test_netcdf_output_writer.jl:889 [inlined]
 [4] macro expansion
   @ ~/.julia/juliaup/julia-1.10.2+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
 [5] top-level scope
   @ ~/github/Oceananigans.jl/test/test_netcdf_output_writer.jl:880
NetCDF output writer [CPU]: Test Failed at /Users/jmtzmo/github/Oceananigans.jl/test/test_netcdf_output_writer.jl:727
  Expression: all(isapprox.((ds["c2"])[:, n + 1], c̄2(averaging_times), rtol = rtol))
```

File: src/OutputWriters/netcdf_output_writer.jl [hunk 1]
```diff
@@ -337,8 +349,10 @@ function NetCDFOutputWriter(model, outputs; filename, schedule,
                                    dimensions = Dict(),
                            overwrite_existing = nothing,
                                  deflatelevel = 0,
+                                     previous = 1,
+                                 max_filesize = Inf,
                                       verbose = false)
-
+
```
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:355]:
```suggestion

```

File: src/OutputWriters/netcdf_output_writer.jl [hunk 2]
```diff
@@ -496,6 +457,10 @@ Writes output to netcdf file `output_writer.filepath` at specified intervals. In
 every time an output is written to the file.
 """
 function write_output!(ow::NetCDFOutputWriter, model)
+    # TODO allow user to split by number of snapshots, rathern than filesize.
+    # Start a new file if the filesize exceeds max_filesize
+    filesize(ow.filepath) >= ow.max_filesize && start_next_file(model, ow)
```
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:462]:
```suggestion
    filesize(ow.filepath) ≥ ow.max_filesize && start_next_file(model, ow)
```

File: src/OutputWriters/netcdf_output_writer.jl [hunk 3]
```diff
@@ -583,3 +548,88 @@ dictify(outputs::LagrangianParticles) = Dict("particles" => outputs)
...
+    ow.filepath = replace(ow.filepath, r"part\d+.nc$" => "part" * string(ow.previous) * ".nc")
+    ow.overwrite_existing && isfile(ow.filepath) && rm(ow.filepath, force=true)
+    verbose && @info "Now writing to: $(ow.filepath)"
+
+    initialize_nc_file!(ow, model)
+    
+    return nothing
+end
+
+function initialize_nc_file!(filepath, outputs, schedule, array_type, indices, with_halos, global_attributes, output_attributes, dimensions, overwrite_existing, deflatelevel, model)
```
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:576]:
```suggestion
function initialize_nc_file!(filepath,
                             outputs,
                             schedule,
                             array_type,
                             indices,
                             with_halos,
                             global_attributes,
                             output_attributes,
                             dimensions,
                             overwrite_existing,
                             deflatelevel,
                             model)
```

File: src/OutputWriters/netcdf_output_writer.jl [hunk 4]
```diff
@@ -583,3 +548,88 @@ dictify(outputs::LagrangianParticles) = Dict("particles" => outputs)
...
+
+        sync(dataset)
+    end
+
+    close(dataset)
+    return dataset, outputs
+end
+
+initialize_nc_file!(ow::NetCDFOutputWriter, model) =
+    initialize_nc_file!(ow.filepath, ow.outputs, ow.schedule, ow.array_type, ow.indices, ow.with_halos, ow.global_attributes, ow.output_attributes, ow.dimensions, ow.overwrite_existing, ow.deflatelevel, model)
```
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:635]:
```suggestion
initialize_nc_file!(ow::NetCDFOutputWriter, model) =
    initialize_nc_file!(ow.filepath,
                        ow.outputs,
                        ow.schedule,
                        ow.array_type,
                        ow.indices,
                        ow.with_halos,
                        ow.global_attributes,
                        ow.output_attributes,
                        ow.dimensions,
                        ow.overwrite_existing,
                        ow.deflatelevel,
                        model)

```

File: test/test_netcdf_output_writer.jl [hunk 5]
```diff
@@ -45,6 +45,50 @@ function test_DateTime_netcdf_output(arch)
...
+
+    push!(simulation.output_writers, ow)
+
+    # 531 KiB of output will be written which should get split into 3 files.
+    run!(simulation)
+
+    # Test that files has been split according to size as expected.
+    @test filesize("test_part1.nc") > 200KiB
+    @test filesize("test_part2.nc") > 200KiB
+    @test filesize("test_part3.nc") < 200KiB
```
Comment (navidcy) [inline test/test_netcdf_output_writer.jl:73]:
```suggestion
    @test filesize("test_part1.nc") > max_filesize
    @test filesize("test_part2.nc") > max_filesize
    @test filesize("test_part3.nc") < max_filesize
```

File: src/OutputWriters/netcdf_output_writer.jl [hunk 6]
```diff
@@ -583,3 +561,111 @@ dictify(outputs::LagrangianParticles) = Dict("particles" => outputs)
...
+        end
+
+        sync(dataset)
+    end
+
+    close(dataset)
+    return dataset, outputs
+end
+
+initialize_nc_file!(ow::NetCDFOutputWriter, model) =
```
Comment (navidcy) [inline src/OutputWriters/netcdf_output_writer.jl:658]:
```suggestion
```

File: test/test_netcdf_output_writer.jl [hunk 7]
```diff
@@ -45,6 +45,52 @@ function test_DateTime_netcdf_output(arch)
...
+
+    push!(simulation.output_writers, ow)
+
+    # 531 KiB of output will be written which should get split into 3 files.
+    run!(simulation)
+
+    # Test that files has been split according to size as expected.
+    @test filesize("test_part1.nc") > max_filesize
+    @test filesize("test_part2.nc") > max_filesize
+    @test filesize("test_part3.nc") < max_filesize
```
Comment (navidcy) [inline test/test_netcdf_output_writer.jl:75]:
```suggestion
    @test filesize("test_part3.nc") < max_filesize
    @test !isfile("test_part4.nc")
```

**Comment by glwagner** [comment]
Cool! Very nice!

**Review by navidcy — APPROVED** [review]

**Comment by navidcy** [comment]
@Sbozzolo, why is the distributed CI not starting here? Because this PR is from a fork?

**Comment by Sbozzolo** [comment]
The job is not showing up in the buildkite pipeline at all... 

It should build PRs, can you try commiting again to see if it picked up?

**Comment by glwagner** [comment]
We also had this problem on `Thermodynamics` and it was fixed by re-opening the PR from the repo rather than a fork. There is a security issue because you may not want to run buildkite jobs from random people (despite that we definitely do want to run PRs from @josuemtzmo <3)

**Comment by Sbozzolo** [comment]
Now it is.

(But I can turn it off if you prefer)

**Comment by navidcy** [comment]
hm... OK.

@josuemtzmo any chance you can re-open this as a branch of the main repo? we have the convention that branches are named with `initials/my-branch-name` (dashes), e.g., `ncc/the-best-new-feature`. I invited you as a collaborator in the repo.

**Comment by josuemtzmo** [comment]
Ok, I will push to the `upstream`, I'm closing this PR in preference to that other one.

**Comment by josuemtzmo** [comment]
This PR was replaced by #3512

**Comment by glwagner** [comment]
> Now it is.
> 
> (But I can turn it off if you prefer)

Well it is enabled for Oceananigans too. So I guess there is a different reason that forks are problematic.

**Pull request closed without merging**